plugins/catalog-backend/src/ingestion/processors/PlaceholderProcessor.ts

```diff
@@ -0,0 +1,214 @@
+        // We're an object but no placeholders at this level - process all
+        // entries recursively
+        const entries = await Promise.all(
+          Object.entries(data).map(([k, v]) =>
+            process(v).then(vp => [k, vp] as const),
+          ),
+        );
+        return entries.every(([, [, changed]]) => !changed)
+          ? [data, false]
+          : [Object.fromEntries(entries.map(([k, [v]]) => [k, v])), true];
+      } else if (keys.length !== 1) {
+        throw new Error(
+          'Placeholders have to be on the form of a single $-prefixed key in an object',
+        );
+      }
+
+      const resolverKey = keys[0].substr(1);
+      const resolver = this.resolvers[resolverKey];
+      if (!resolver) {
+        throw new Error(`Encountered unknown placeholder \$${resolverKey}`);
+      }
+
+      return [
+        await resolver({
+          key: resolverKey,
+          value: data[keys[0]],
+          location,
+          read,
+        }),
+        true,
+      ];
+    };
+
+    const [result] = await process(entity);
+    return result;
+  }
+}
+
+/*
+ * Resolvers
+ */
+
+export async function yamlPlaceholderResolver(
+  params: ResolverParams,
+): Promise<JsonValue> {
+  const text = await readTextLocation(params);
+
+  let documents: yaml.Document.Parsed[];
+  try {
+    documents = yaml.parseAllDocuments(text).filter(d => d);
+  } catch (e) {
+    throw new Error(
+      `Placeholder \$${params.key} failed to parse YAML data at ${params.value}, ${e}`,
+    );
+  }
+
+  if (documents.length !== 1) {
+    throw new Error(
+      `Placeholder \$${params.key} expected to find exactly one document of data at ${params.value}, found ${documents.length}`,
+    );
+  }
+
+  const document = documents[0];
+
+  if (document.errors?.length) {
+    throw new Error(
+      `Placeholder \$${params.key} found an error in the data at ${params.value}, ${document.errors[0]}`,
+    );
+  }
+
+  return document.toJSON();
+}
+
+export async function jsonPlaceholderResolver(
+  params: ResolverParams,
+): Promise<JsonValue> {
+  const text = await readTextLocation(params);
+
+  try {
+    return JSON.parse(text);
+  } catch (e) {
+    throw new Error(
+      `Placeholder \$${params.key} failed to parse JSON data at ${params.value}, ${e}`,
+    );
+  }
+}
+
+export async function textPlaceholderResolver(
+  params: ResolverParams,
+): Promise<JsonValue> {
+  return await readTextLocation(params);
+}
+
+/*
+ * Helpers
+ */
+
+async function readTextLocation(params: ResolverParams): Promise<string> {
+  const newLocation = relativeLocation(params);
+
+  try {
+    const response = await params.read(newLocation);
+    if (response.type !== 'data') {
+      throw new Error(`Expected data, got ${response.type}`);
+    }
+    return response.data.toString('utf-8');
+  } catch (e) {
+    throw new Error(
+      `Placeholder \$${params.key} could not read location ${params.value}, ${e}`,
+    );
+  }
+}
+
+function relativeLocation({
```

Comment:
@freben I have a short question here, do I understand it correctly that right now only relative URLs are supported? But I would be open to hand in a PR for that.

Reply:
But I guess that would require to get rid of the location `type` first 😉

Reply:
No! You can put an absolute URL here too. It gets passed to the two-argument `URL` constructor which deals with either form. But it has to be to the same `type` of location "just because reasons", until we get rid of that thing...

Reply:
Hmm, I get 

```
[1] 2020-10-02T09:15:57.923Z catalog warn Processor PlaceholderProcessor threw an error while processing entity API:default/spotify at file ../catalog-model/examples/spotify-api.yaml, Error: Placeholder $text could not form an URL out of ../catalog-model/examples/spotify-api.yaml and https://raw.githubusercontent.com/APIs-guru/openapi-directory/master/APIs/spotify.com/v1/swagger.yaml
```

But maybe the problem is, that my first location is relative and the second absolute 😆

Reply:
Ok, odd! The constructor works with those arguments in the dev console. I wonder if the node12 URL is somehow different and worse than the browser one ........

Reply:
Oh hang on, those were in the reverse order so when I put them in the other order in the URL constructor it indeed did throw.

Now the question is, how did the actual input location turn out to be a relative path? The processor chain encountered an entity that was coming from `../catalog-model/examples/spotify-api.yaml`; that makes me think that you are using the `file` protocol for this right? That probably isn't supported in conjunction with these placeholders.

I wonder how to best do this. Would you like to contribute support for this case to the code in PlaceholderProcessor.ts?

Maybe some special case to check that if location.target is NOT parseable as a URL (the constructor throws), but the other one IS parseable, then return the other one. I mean if we wan to support the other one AS WELL not being parseable as a URL, then we will have to have some kind of file system traversal fallback thing and I'm not sure we want to do that anyway (just include `$file: /etc/shadow` or whatever)

Reply:
> ... are using the file protocol for this right?
Yeah, for testing locally 😉 I don't need it in production. But I guess we should fix it anyway.

Yeah, I can think about PR with your suggestion, thanks for checking this!

Reply:
@freben  I wonder if one can pass an absolute `file://` url into Backstage to read data from anywhere on the host 🤔

Reply:
when we remove the type, yeah maybe, haven't thought that far yet